[Kernel] Speed up fused MoE LoRA triton op

cwazai opened the pull request:
#Background / Motivation
_fused_moe_lora uses a Triton kernel to compute MoE-routed LoRA GEMMs (shrink + expand). In the current implementation, the kernel and host launcher have a few overheads that become noticeable for large token counts / many LoRA adapters:

extra int64 arithmetic and modulo ops in the hot path,
repeated bounds checks using -1 sentinels with int64 comparisons,
expand path materializes an intermediate buffer (b_intermediate_cache1) and then loops in Python to accumulate into output,
a_intermediate_cache1 is allocated every call (and zero-filled unconditionally).
This PR focuses on reducing kernel-side overhead and avoiding redundant memory traffic/allocations while keeping the functional behavior unchanged.

#What’s changed
1) Kernel: cheaper bounds checks + more int32/uint32 math
Added MAX_LORAS_TOTAL as a scalar kernel arg and included it in do_not_specialize.
Converted several scalars to tl.uint32/tl.int32 to reduce int64 ops.
Replaced lora_id == -1 sentinel check with an unsigned bounds check:
lora_id loaded as int32, cast to uint32; (-1) becomes 0xFFFF_FFFF and fails < MAX_LORAS_TOTAL.
Similarly validated expert_id via uint32 < num_experts.
Benefit: fewer branches and cheaper comparisons; avoids subtle sentinel-dependent behavior.

2) Kernel: remove % N and use mask for N dimension
Replaced (… % N) addressing for offs_bn with direct offs_cn and cn_mask = offs_cn < N.
This avoids modulo in the inner addressing and keeps correctness via masking.
3) Kernel: optional B-side L2 cache modifier
Added USE_B_L2_CACHE (constexpr) and uses cache_modifier=".ca" for B loads when enabled.
4) Kernel: support ADD_TO_C to accumulate directly into output
Added ADD_TO_C (constexpr):
when SPLIT_K == 1 and ADD_TO_C=True, the kernel loads previous C and stores (acc + prev) (float32 accumulation).
otherwise keeps existing store/atomic_add behavior.
Benefit: enables expand path to accumulate into output directly, removing a separate intermediate output buffer + Python-side loop.

## Testing
```shell
**unit tests** (`tests/lora/test_fused_moe_lora_kernel.py`): All pass
[befoe]
======================================================================================== 48 passed, 14 warnings in 279.58s (0:04:39) ========================================================================================
[after]

=============================================================== 48 passed, 14 warnings in 146.20s (0:02:26) ===============================================================
```


```python

## Benchmarks
Hardware: Nvidia H800, Software: CUDA/driver：13.0
Workload: model：Qwen3-30B-A3B, rank：64

50%base model/50% lora（Three Lora）
[before] 
============ Serving Benchmark Result ============
Successful requests:                     800       
Failed requests:                         0         
Maximum request concurrency:             80        
Benchmark duration (s):                  33.83     
Total input tokens:                      819200    
Total generated tokens:                  40000     
Request throughput (req/s):              23.65     
Output token throughput (tok/s):         1182.45   
Peak output token throughput (tok/s):    2875.00   
Peak concurrent requests:                115.00    
Total Token throughput (tok/s):          25398.93  
---------------Time to First Token----------------
Mean TTFT (ms):                          444.21    
Median TTFT (ms):                        307.73    
P99 TTFT (ms):                           2366.87   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          59.59     
Median TPOT (ms):                        61.83     
P99 TPOT (ms):                           64.91     
---------------Inter-token Latency----------------
Mean ITL (ms):                           58.39     
Median ITL (ms):                         24.29     
P99 ITL (ms):                            164.44    
==================================================
[after]
============ Serving Benchmark Result ============
Successful requests:                     800       
Failed requests:                         0         
Maximum request concurrency:             80        
Benchmark duration (s):                  31.02     
Total input tokens:                      819200    
Total generated tokens:                  40000     
Request throughput (req/s):              25.79     
Output token throughput (tok/s):         1289.28   
Peak output token throughput (tok/s):    3157.00   
Peak concurrent requests:                119.00    
Total Token throughput (tok/s):          27693.80  
---------------Time to First Token----------------
Mean TTFT (ms):                          464.62    
Median TTFT (ms):                        401.12    
P99 TTFT (ms):                           2216.33   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          53.47     
Median TPOT (ms):                        54.69     
P99 TPOT (ms):                           59.97     
---------------Inter-token Latency----------------
Mean ITL (ms):                           52.40     
Median ITL (ms):                         23.42     
P99 ITL (ms):                            151.51    
==================================================

```


Single Lora


```python
[before] 
============ Serving Benchmark Result ============
Successful requests:                     800       
Failed requests:                         0         
Maximum request concurrency:             80        
Benchmark duration (s):                  36.22     
Total input tokens:                      819200    
Total generated tokens:                  40000     
Request throughput (req/s):              22.09     
Output token throughput (tok/s):         1104.42   
Peak output token throughput (tok/s):    3074.00   
Peak concurrent requests:                114.00    
Total Token throughput (tok/s):          23722.91  
---------------Time to First Token----------------
Mean TTFT (ms):                          520.80    
Median TTFT (ms):                        488.09    
P99 TTFT (ms):                           2617.83   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          62.91     
Median TPOT (ms):                        63.72     
P99 TPOT (ms):                           69.72     
---------------Inter-token Latency----------------
Mean ITL (ms):                           61.65     
Median ITL (ms):                         23.10     
P99 ITL (ms):                            168.25    
==================================================

[after]
============ Serving Benchmark Result ============
Successful requests:                     800       
Failed requests:                         0         
Maximum request concurrency:             80        
Benchmark duration (s):                  33.53     
Total input tokens:                      819200    
Total generated tokens:                  40000     
Request throughput (req/s):              23.86     
Output token throughput (tok/s):         1193.11   
Peak output token throughput (tok/s):    2883.00   
Peak concurrent requests:                115.00    
Total Token throughput (tok/s):          25628.07  
---------------Time to First Token----------------
Mean TTFT (ms):                          520.28    
Median TTFT (ms):                        451.36    
P99 TTFT (ms):                           2402.89   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          57.44     
Median TPOT (ms):                        58.67     
P99 TPOT (ms):                           64.81     
---------------Inter-token Latency----------------
Mean ITL (ms):                           56.29     
Median ITL (ms):                         21.94     
P99 ITL (ms):                            154.61    
==================================================
```

Five LoRa

```python
[before] 
============ Serving Benchmark Result ============
Successful requests:                     800       
Failed requests:                         0         
Maximum request concurrency:             80        
Benchmark duration (s):                  42.43     
Total input tokens:                      819200    
Total generated tokens:                  40000     
Request throughput (req/s):              18.86     
Output token throughput (tok/s):         942.78    
Peak output token throughput (tok/s):    2634.00   
Peak concurrent requests:                110.00    
Total Token throughput (tok/s):          20250.97  
---------------Time to First Token----------------
Mean TTFT (ms):                          515.25    
Median TTFT (ms):                        373.09    
P99 TTFT (ms):                           2954.47   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          75.60     
Median TPOT (ms):                        77.43     
P99 TPOT (ms):                           82.48     
---------------Inter-token Latency----------------
Mean ITL (ms):                           74.09     
Median ITL (ms):                         31.31     
P99 ITL (ms):                            190.65    
==================================================

[after]
============ Serving Benchmark Result ============
Successful requests:                     800       
Failed requests:                         0         
Maximum request concurrency:             80        
Benchmark duration (s):                  39.50     
Total input tokens:                      819200    
Total generated tokens:                  40000     
Request throughput (req/s):              20.25     
Output token throughput (tok/s):         1012.64   
Peak output token throughput (tok/s):    2683.00   
Peak concurrent requests:                111.00    
Total Token throughput (tok/s):          21751.52  
---------------Time to First Token----------------
Mean TTFT (ms):                          495.14    
Median TTFT (ms):                        351.05    
P99 TTFT (ms):                           2734.17   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          70.07     
Median TPOT (ms):                        71.80     
P99 TPOT (ms):                           76.75     
---------------Inter-token Latency----------------
Mean ITL (ms):                           68.67     
Median ITL (ms):                         28.99     
P99 ITL (ms):                            178.76    
==================================================
```
----------------------------------------------------------------------------------------

cwazai committed:
1.[Kernel] Optimize fused MoE LoRA triton op (in-kernel accumulation + workspace reuse)
2.[Kernel] Add correctness + cudagraph tests for fused MoE LoRA op

cwazai committed:
fix CI

cwazai committed:
Merge branch 'main' into fused-moe-lora-opt

LucasWilkinson commented:
can we use `vllm/v1/worker/workspace.py` for the workspace management?

cwazai committed:
fix CI

cwazai committed:
Merge branch 'fused-moe-lora-opt' of https://github.com/cwazai/vllm into fused-moe-lora-opt

cwazai committed:
Fix: Enable Dual-Stream LoRA Computation by Resolving Base-LoRA Dependency in expandKernel

cwazai committed:
Add a prefill/TTFT fast-path: when not capturing CUDA graphs, directly accumulate LoRA results into the output (ADD_TO_C) to reduce TTFT.

cwazai commented on vllm/lora/ops/triton_ops/fused_moe_lora_op.py:
```diff
@@ -368,43 +391,48 @@ def _fused_moe_lora_expand(
+        num_warps=num_warps,
+        num_stages=num_stages,
     )
-    for i in range(num_slices):
```
Thank you for your previous feedback. It was a valuable insight.
I've addressed the concern regarding the dependency between the base model and LoRA computation blocking the dual streams. The issue has been fixed. I've also updated the PR description and supplemented it with the relevant test data/results.
Could you please help to review the changes when you have time? If any further modifications or additional tests are needed, I'm more than happy to follow up promptly.
Thanks again for your guidance and time!

cwazai committed:
Fix OOB in fused MoE-LoRA triton kernel for padded token ids; fuse expand accumulation & reduce overhead

cwazai commented:
> Could you please fix the CI failure first?

Thank you for your previous feedback on the dependency issue. I've revised the code to resolve the conflict between the base model and LoRA computation, ensuring that the dual streams mechanism can now function properly.
I've also supplemented the performance data under concurrent loads involving both the base model and LoRA. The results demonstrate a clear performance improvement when utilizing dual streams for LoRA compared to the previous approach.
Could you please review the updated code and data when you have a moment? I'm ready to make any further adjustments or run additional tests as needed.
Thanks again for your time and guidance!

cwazai committed:
fix(merge): resolve conflicts with main branch

cwazai commented:
> @cwazai Sorry for missing this PR, could you please fix the conflict first, I wll look at this PR ASAP
Hello maintainers! 
I wanted to kindly follow up on this PR. It's quite important for me because it serves as a key deliverable and summary of my recent work efforts. I've invested significant time and energy into it.
Could you please take a look when you have a moment? I'm happy to provide more information or make any necessary changes​ to meet the requirements.
Thank you for your time and consideration!

gnovack commented on vllm/lora/ops/triton_ops/fused_moe_lora_op.py:
```diff
@@ -286,27 +317,59 @@ def _fused_moe_lora_shrink(
+    key = (device.type, device.index, dtype, num_slices, M, top_k_num, rank)
+    buf = _A_WS_CACHE.get(key)
+    if buf is None:
+        buf = torch.empty((num_slices, M, top_k_num, rank), device=device, dtype=dtype)
```
Have you run the e2e lora tests (e.g. `tests/lora/test_olmoe_tp.py`) to validate that initializing the cache as empty does not cause any accuracy issues? 

From what I recall, we needed to initialize this cache with zeros since some threads in `_fused_moe_lora_kernel` exit before writing anything to the output buffer (e.g. when `expert_id == -1` or `lora_id == -1`) leading to stale data being read

cwazai replied:
Yes, the e2e LoRA tests (including test_olmoe_tp.py) pass successfully with the latest code. The results confirm that initializing the cache as empty is safe and maintains numerical accuracy, as the kernel now properly handles early-exit threads.

gnovack commented:
One high-level question - do you have any benchmarking/profiling data to quantify the performance gains of the each of the first 3 changes independently? I'm curious how much of the end-to-end gains you are seeing are attributable to these changes.

i.e.
1. cheaper bounds checks + more int32/uint32 math
2. remove % N and use mask for N dimension
3. optional B-side L2 cache modifier

cwazai committed:
1. **Kernel optimization**:
   - Remove `MAX_LORAS_TOTAL` from `do_not_specialize` for better Triton specialization
   - Fix numerical consistency in `ADD_TO_C` branch by performing addition in output dtype
   - Add L2 cache hints (`USE_B_L2_CACHE`) for better memory access patterns

2. **Split-K handling**:
   - For `split_k == 1`: Directly accumulate results in output (`ADD_TO_C=True`)
   - For `split_k > 1`: Use atomic addition to zero-initialized workspace, then perform single `add_` to output
   - Maintains numerical equivalence with upstream while preserving parallelism

3. **Memory optimization**:
   - Replace per-M workspace dictionary with single growable buffers per `(device, dtype, num_slices, top_k)`
   - Buffers expand using power-of-two sizing to minimize reallocation frequency
   - Significantly reduces GPU memory fragmentation during dynamic batching

4. **Stride-aware kernel invocation**:
   - Properly handle non-contiguous workspace views from buffer slicing
   - Use `stride(0)` for slice spacing instead of `numel() // num_slices`
   - Fix `view()` compatibility issues with padded workspaces

5. **TP compatibility**:
   - Add explicit `.contiguous()` before tensor model parallel collectives
   - Ensure proper handling when `fully_sharded=True`

cwazai committed:
fix CI

cwazai committed:
Merge branch 'main' into fused-moe-lora-opt

Signed-off-by: cwazai <38356712+cwazai@users.noreply.github.com>
